### DISCUSSION TITLE
nx monorepo with project specific edge functions

### DISCUSSION THREAD

**Discussion by th-m** [vercel/community#916] [Help] [answered]
I have an NX monorepo with a single `vercel.json` file at the root. Which seems problematic to me, I wish vercel would allow configurable paths to `vercel.json` so I could no exactly what project is getting what... but I digress.

`vercel.json`

```
{
  "rewrites": [
    {
      "source": "/(.*)",
      "destination": "/"
    }
  ],
   "github": {
    "autoJobCancelation": false
  }
}
```

This walkthrough describes how I have my projects configured https://nx.dev/recipe/deploy-nextjs-to-vercel

https://vercel.com/docs/concepts/functions/serverless-functions

The vercel docs say 

> To deploy a serverless Node.js API, create a function in a .js file within the /api directory at the root of your project:

However, since I am in a monorepo with all projects sharing the same root I do not want my edge functions duplicated across every project. I want to be able to specify which projects should have which edge functions associated to them. Is this possible? Am I missing some config tool that could allow this? 

Thanks in advance for the help.

**Comment by amyegan** [comment]
Hi @th-m! If you're using Next.js, I would recommend using [`next.config.js`](https://nextjs.org/docs/api-reference/next.config.js/introduction) to configure [rewrites](https://nextjs.org/docs/api-reference/next.config.js/rewrites). The desire for more specific control in `vercel.json` is noted. 

The Serverless Functions `/api` directory belongs at the app root. For example, I have a minimal app in my test Nx monorepo with this file structure:
<img width="126" alt="image" src="https://user-images.githubusercontent.com/8247105/193365266-15f01998-06fc-4200-a85a-23ef4c7d23a5.png">

You would do almost the same thing to [create an edge function](https://vercel.com/docs/concepts/functions/edge-functions#creating-edge-functions).

**Reply by th-m** [reply]
Yeah the API/ root path is not picking up edge functions in my react apps

**Comment by th-m** [comment]
Yeah the structure you are showing there is what I have done for my nx
apps, but none of the edge functions are showing up or registering.

**Reply by amyegan** [reply]
If you can share a [minimal reproducible example](https://vercel.com/support/articles/creating-a-minimal-reproducible-example) and example deployment, then I can take a closer look to find what is going wrong.

**Comment by th-m** [comment, marked as answer]
I actually figured it out.

Because I am using NX my "projects" share the same root in the repository.
[image: Screen Shot 2022-10-03 at 4.52.48 PM.png]

I was able to get some edge functions deploying like this.
However, though this is a valid way to get edge functions push out, it
deploys the functions to every project, which is not ideal.
Also, I discovered that the build-ignore check will cancel edge function
deploys. Which I thought was silly because the whole point of edge
functions is that they exist outside of the build and are built
automagically.
What is more, the build-ignore check will remove any pre-existing functions
related to a project. This "feature" was particularly frustrating.

What I would recommend for you guys.

Make the `vercel.json` configurable right with the build options in the UI.
Allow the edge functions path to be configurable also either preferable in
the `vercel.json`

Note in your docs that some features do not work as documented when using
NX monorepo. e.g. `vercel.json` and `serverless functions` are defined at
the root of the repo and applied to all projects.

Also in the vercel docs regarding serverless functions. Please show
configuration steps for projects that are not next.js I got the impression
that those examples were the standard when in fact it is more the exception
being that they utilize `next.config` which is something vercel doesn't
afford to other tech stacks.

Love what you guys are doing. I will admit that was one of the most
frustrating experiences trying to deduce what was happening inside your
black box, even still, you guys provide a huge strategic advantage to us
and our product and so I am grateful.

**Reply by amyegan** [reply]
Thanks for sharing this feedback. I'm glad you still love Vercel, and it's also very important for us to hear about what is _not_ going well so we know what to improve! I'm sharing this feedback with the team so they can better understand the problems you're facing and guide future development. If we're lucky, they may even have some suggestions for things to try that I'm not yet aware of

**Comment by th-m** [comment]
Thank you
